File: leetcode/101-200/148-sort-list/bench/sort_list.py
"""Benchmark workload for LeetCode #148 — Sort List (Python; scale lane)."""

import sys


def split_mid(nxt, head):
    slow = head
    fast = nxt[head]
    while fast != -1:
        fast = nxt[fast]
        if fast != -1:
            slow = nxt[slow]
            fast = nxt[fast]
    mid = nxt[slow]
    nxt[slow] = -1
    return mid


def merge(val, nxt, a, b):
    ai, bi = a, b
    head = -1
    tail = -1
    while ai != -1 and bi != -1:
        if val[ai] <= val[bi]:
            if head == -1:
                head = ai
            else:
                nxt[tail] = ai
            tail = ai
            ai = nxt[ai]
        else:
            if head == -1:
                head = bi
            else:
                nxt[tail] = bi
            tail = bi
            bi = nxt[bi]
    rest = ai if ai != -1 else bi
    if tail != -1:
        if rest == -1:
            nxt[tail] = -1
        else:
            nxt[tail] = rest
    return head


def sort_chain(val, nxt, head):
    if head == -1:
        return -1
    if nxt[head] == -1:
        return head
    mid = split_mid(nxt, head)
    left = sort_chain(val, nxt, head)
    right = sort_chain(val, nxt, mid)
    return merge(val, nxt, left, right)


def main():
    sys.setrecursionlimit(1000000)
    n = 20000
    passes = 180
    vr = 100000
    val = [0] * n
    nxt = [-1] * n
    state = 12345
    for i in range(n):
        state = (state * 1103515245 + 12345) & 2147483647
        val[i] = state % vr

    sink = 0
    for _p in range(passes):
        state = (state * 1103515245 + 12345) & 2147483647
        idx = state % n
        state = (state * 1103515245 + 12345) & 2147483647
        val[idx] = state % vr

        for i in range(n - 1):
            nxt[i] = i + 1
        nxt[n - 1] = -1

        head = sort_chain(val, nxt, 0)

        cur = head
        pos = 1
        while cur != -1:
            sink += pos * val[cur]
            pos += 1
            cur = nxt[cur]
    print(sink)


main()

File: leetcode/101-200/148-sort-list/bench/sort_list.rs
fn split_mid(nxt: &mut [i64], head: i64) -> i64 {
    let mut slow = head;
    let mut fast = nxt[head as usize];
    while fast != -1 {
        fast = nxt[fast as usize];
        if fast != -1 {
            slow = nxt[slow as usize];
            fast = nxt[fast as usize];
        }
    }
    let mid = nxt[slow as usize];
    nxt[slow as usize] = -1;
    mid
}

fn merge(val: &[i64], nxt: &mut [i64], a: i64, b: i64) -> i64 {
    let mut ai = a;
    let mut bi = b;
    let mut head: i64 = -1;
    let mut tail: i64 = -1;
    while ai != -1 && bi != -1 {
        if val[ai as usize] <= val[bi as usize] {
            if head == -1 { head = ai; } else { nxt[tail as usize] = ai; }
            tail = ai;
            ai = nxt[ai as usize];
        } else {
            if head == -1 { head = bi; } else { nxt[tail as usize] = bi; }
            tail = bi;
            bi = nxt[bi as usize];
        }
    }
    let rest = if ai != -1 { ai } else { bi };
    if tail != -1 {
        if rest == -1 { nxt[tail as usize] = -1; } else { nxt[tail as usize] = rest; }
    }
    head
}

fn sort_chain(val: &[i64], nxt: &mut [i64], head: i64) -> i64 {
    if head == -1 { return -1; }
    if nxt[head as usize] == -1 { return head; }
    let mid = split_mid(nxt, head);
    let left = sort_chain(val, nxt, head);
    let right = sort_chain(val, nxt, mid);
    merge(val, nxt, left, right)
}

fn main() {
    let n: i64 = 20000;
    let passes: i64 = 180;
    let vr: i64 = 100000;

    let mut val = vec![0i64; n as usize];
    let mut nxt = vec![-1i64; n as usize];
    let mut state: i64 = 12345;
    for i in 0..n as usize {
        state = (state * 1103515245 + 12345) & 2147483647;
        val[i] = state % vr;
    }

    let mut sink: i64 = 0;
    for _p in 0..passes {
        state = (state * 1103515245 + 12345) & 2147483647;
        let idx = (state % n) as usize;
        state = (state * 1103515245 + 12345) & 2147483647;
        val[idx] = state % vr;

        for i in 0..(n - 1) as usize {
            nxt[i] = i as i64 + 1;
        }
        nxt[(n - 1) as usize] = -1;

        let head = sort_chain(&val, &mut nxt, 0);

        let mut cur = head;
        let mut pos: i64 = 1;
        while cur != -1 {
            sink += pos * val[cur as usize];
            pos += 1;
            cur = nxt[cur as usize];
        }
    }
    println!("{}", sink);
}
